File: learnLeetCode_remote/Day24/Part5/truth_chain.py
# The Truth Chain
# Master of Truth: trusts nobody (outdegree 0) and is trusted (transitively) by everyone else.
# Approach: build reverse graph; candidate must have outdeg 0 in original, and from candidate in reversed graph we can reach all nodes.
from typing import List, Tuple
from collections import defaultdict, deque

def master_of_truth(n: int, trust: List[Tuple[int,int]]) -> int:
    g = [[] for _ in range(n)]
    gr = [[] for _ in range(n)]
    outdeg = [0]*n
    for a,b in trust:
        a-=1; b-=1
        g[a].append(b)
        gr[b].append(a)
        outdeg[a]+=1
    # candidates: outdeg==0
    cand = [i for i in range(n) if outdeg[i]==0]
    for c in cand:
        # BFS from c in reversed graph -> nodes that can reach c in original
        vis = [False]*n
        dq = deque([c])
        vis[c]=True
        seen = 1
        while dq:
            u = dq.popleft()
            for v in gr[u]:
                if not vis[v]:
                    vis[v]=True; seen+=1; dq.append(v)
        if seen==n:
            return c+1
    return -1

if __name__ == "__main__":
    n=4; trust=[(1,3),(2,3),(3,4)]
    print(master_of_truth(n, trust))  # 4

File: learnLeetCode_remote/Day24/Part5/truth_chain.cpp
#include <bits/stdc++.h>
using namespace std;

// The Truth Chain
// Candidate must have outdegree 0 and be reachable (in reverse graph) from all others.

int master_of_truth(int n, const vector<pair<int,int>>& trust){
    vector<vector<int>> gr(n); vector<int> outdeg(n,0);
    for(auto [a,b]: trust){ --a; --b; gr[b].push_back(a); outdeg[a]++; }
    vector<int> cand; for(int i=0;i<n;++i) if(outdeg[i]==0) cand.push_back(i);
    for(int c: cand){
        vector<char> vis(n,0); queue<int> q; q.push(c); vis[c]=1; int seen=1;
        while(!q.empty()){
            int u=q.front(); q.pop();
            for(int v: gr[u]) if(!vis[v]){ vis[v]=1; ++seen; q.push(v); }
        }
        if(seen==n) return c+1;
    }
    return -1;
}

int main(){
    int n=4; vector<pair<int,int>> trust={{1,3},{2,3},{3,4}};
    cout << master_of_truth(n, trust) << "\n"; // 4
    return 0;
}
